File: Rosella/src/main/java/me/hydos/rosella/render/vertex/BufferVertexConsumer.java
package me.hydos.rosella.render.vertex;

import me.hydos.rosella.device.VulkanDevice;
import me.hydos.rosella.memory.Memory;

import java.nio.ByteBuffer;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

// FIXME redo this whole class honestly
//public final class BufferVertexConsumer {
//    private final VertexFormat format;
//
//    private ByteBuffer buffer;
//    private int vertexCount;
//    private int debugSize;
//
//    public BufferVertexConsumer(VertexFormat format) {
//        this.format = format;
//        this.buffer = null;
//    }
//
//    public List<ManagedBuffer> getBuffers() {
//        return Collections.singletonList(new ManagedBuffer(buffer, 0, 0, vertexCount + format.getSize(), false));
//    }
//
//    public VertexFormat getFormat() {
//        return format;
//    }
//
//    public BufferVertexConsumer pos(final float x, final float y, final float z) {
//        this.buffer.putFloat(x);
//        this.buffer.putFloat(y);
//        this.buffer.putFloat(z);
//        this.debugSize += 12;
//        return this;
//    }
//
//    public BufferVertexConsumer color(final byte red, final byte green, final byte blue) {
//        this.buffer.put(red);
//        this.buffer.put(green);
//        this.buffer.put(blue);
//        this.debugSize += 3;
//        return this;
//    }
//
//    public BufferVertexConsumer color(final byte red, final byte green, final byte blue, final byte alpha) {
//        this.buffer.put(red);
//        this.buffer.put(green);
//        this.buffer.put(blue);
//        this.buffer.put(alpha);
//        this.debugSize += 4;
//        return this;
//    }
//
//    public BufferVertexConsumer normal(final float x, final float y, final float z) {
//        this.buffer.putFloat(x);
//        this.buffer.putFloat(y);
//        this.buffer.putFloat(z);
//        this.debugSize += 12;
//        return this;
//    }
//
//    public BufferVertexConsumer uv(final float u, final float v) {
//        this.buffer.putFloat(u);
//        this.buffer.putFloat(v);
//        this.debugSize += 8;
//        return this;
//    }
//
//    public BufferVertexConsumer uv(final short u, final short v) {
//        this.buffer.putShort(u);
//        this.buffer.putShort(v);
//        this.debugSize += 4;
//        return this;
//    }
//
//    public BufferVertexConsumer putByte(final int index, final byte value) {
//        this.buffer.put(index, value);
//        this.debugSize++;
//        return this;
//    }
//
//    public BufferVertexConsumer putShort(final int index, final short value) {
//        this.buffer.putShort(index, value);
//        this.debugSize += 2;
//        return this;
//    }
//
//    public BufferVertexConsumer putFloat(final int index, final float value) {
//        this.buffer.putFloat(index, value);
//        this.debugSize += 4;
//        return this;
//    }
//
//    public BufferVertexConsumer nextVertex() {
//        if (this.debugSize != this.format.getSize()) {
//            throw new RuntimeException("Incorrect vertex size passed. Received " + this.debugSize + " but wanted " + this.format.getSize());
//        } else {
//            this.debugSize = 0;
//            this.vertexCount++;
//            return this;
//        }
//    }
//
//    public void clear() {
//        this.buffer.clear(); // FIXME make a new buffer
//        this.vertexCount = 0;
//    }
//
//    public void free(VulkanDevice device, Memory memory) {
//        // FIXME
//    }
//
//    public int getVertexSize() {
//        return this.format.getSize();
//    }
//
//    public int getVertexCount() {
//        return this.vertexCount;
//    }
//
//    @Override
//    public boolean equals(Object o) {
//        if (this == o) return true;
//        if (o == null || getClass() != o.getClass()) return false;
//        BufferVertexConsumer that = (BufferVertexConsumer) o;
//        return vertexCount == that.vertexCount && debugSize == that.debugSize && format.equals(that.format) && buffer.equals(that.buffer);
//    }
//
//    @Override
//    public int hashCode() {
//        return Objects.hash(format, buffer, vertexCount, debugSize);
//    }
//}

File: Rosella/src/main/kotlin/me/hydos/rosella/render/vertex/BufferVertexConsumer.kt
package me.hydos.rosella.render.vertex

import java.nio.ByteBuffer
import java.util.*
import java.util.function.Consumer

class BufferVertexConsumer(override val format: VertexFormat) : VertexConsumer {

	var bufferConsumerList : MutableList<Consumer<ByteBuffer>> = ArrayList()
	private var vertexSize = format.getSize()
	private var vertexCount = 0

	private var debugSize = 0

	override fun pos(x: Float, y: Float, z: Float): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putFloat(x)
			it.putFloat(y)
			it.putFloat(z)
		})
		debugSize += 3 * Float.SIZE_BYTES
		return this
	}

	override fun color(red: Byte, green: Byte, blue: Byte): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.put(red)
			it.put(green)
			it.put(blue)
		})

		debugSize += 3 * Byte.SIZE_BYTES
		return this
	}

	override fun color(red: Byte, green: Byte, blue: Byte, alpha: Byte): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.put(red)
			it.put(green)
			it.put(blue)
			it.put(alpha)
		})

		debugSize += 4 * Byte.SIZE_BYTES
		return this
	}

	override fun normal(x: Float, y: Float, z: Float): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putFloat(x)
			it.putFloat(y)
			it.putFloat(z)
		})

		debugSize += 3 * Float.SIZE_BYTES
		return this
	}

	override fun uv(u: Float, v: Float): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putFloat(u)
			it.putFloat(v)
		})
		debugSize += 2 * Float.SIZE_BYTES
		return this
	}

	override fun uv(u: Short, v: Short): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putShort(u)
			it.putShort(v)
		})

		debugSize += 2 * Short.SIZE_BYTES
		return this
	}

	fun putByte(index: Int, value: Byte): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.put(index, value)
		})

		debugSize += Byte.SIZE_BYTES
		return this;
	}

	fun putShort(index: Int, value: Short): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putShort(index, value)
		})

		debugSize += Short.SIZE_BYTES
		return this;
	}

	fun putFloat(index: Int, value: Float): VertexConsumer {
		bufferConsumerList.add(Consumer {
			it.putFloat(index, value)
		})

		debugSize += Float.SIZE_BYTES
		return this;
	}

	override fun nextVertex(): VertexConsumer {
		if (debugSize != vertexSize) {
			throw RuntimeException("Incorrect vertex size passed. Received $debugSize but wanted $vertexSize")
		}

		debugSize = 0
		vertexCount++
		return this
	}

	override fun clear() {
		bufferConsumerList.clear()
		vertexCount = 0
	}

	override fun getVertexSize(): Int {
		return vertexSize
	}

	override fun getVertexCount(): Int {
		return vertexCount
	}

	override fun copy(): VertexConsumer {
		val consumer = BufferVertexConsumer(format)
		consumer.debugSize = this.debugSize
		consumer.bufferConsumerList = Collections.unmodifiableList(this.bufferConsumerList)
		consumer.vertexCount = this.vertexCount
		consumer.vertexSize = this.vertexSize
		return consumer
	}

	override fun hashCode(): Int {
		return Objects.hash(bufferConsumerList, vertexSize, format)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (javaClass != other?.javaClass) return false

		other as BufferVertexConsumer

		if (format != other.format) return false
		if (bufferConsumerList != other.bufferConsumerList) return false
		if (vertexSize != other.vertexSize) return false

		return true
	}

}